interfaces/repo.go

```diff
@@ -553,23 +563,50 @@ func (r *Repository) Connect(ref ConnRef) error {
 	// Ensure that such plug exists
 	plug := r.plugs[plugSnapName][plugName]
 	if plug == nil {
-		return fmt.Errorf("cannot connect plug %q from snap %q, no such plug", plugName, plugSnapName)
+		return nil, fmt.Errorf("cannot connect plug %q from snap %q: no such plug", plugName, plugSnapName)
 	}
 	// Ensure that such slot exists
 	slot := r.slots[slotSnapName][slotName]
 	if slot == nil {
-		return fmt.Errorf("cannot connect plug to slot %q from snap %q, no such slot", slotName, slotSnapName)
+		return nil, fmt.Errorf("cannot connect slot %q from snap %q: no such slot", slotName, slotSnapName)
 	}
 	// Ensure that plug and slot are compatible
 	if slot.Interface != plug.Interface {
-		return fmt.Errorf(`cannot connect plug "%s:%s" (interface %q) to "%s:%s" (interface %q)`,
+		return nil, fmt.Errorf(`cannot connect plug "%s:%s" (interface %q) to "%s:%s" (interface %q)`,
 			plugSnapName, plugName, plug.Interface, slotSnapName, slotName, slot.Interface)
 	}
 	// Ensure that slot and plug are not connected yet
 	if r.slotPlugs[slot][plug] != nil {
 		// But if they are don't treat this as an error.
-		return nil
+		return r.slotPlugs[slot][plug], nil
+	}
+
+	cplug := NewConnectedPlug(plug, plugDynamicAttrs)
+	cslot := NewConnectedSlot(slot, slotDynamicAttrs)
+
+	iface, ok := r.ifaces[plug.Interface]
```

Comment:
Can you move this check above, so that it happens before we construct `NewConnected{Plug,Slot}` please?